flakey-test-tracking/src/github.js

```diff
@@ -0,0 +1,261 @@
+const { Octokit } = require('@octokit/rest')
+const { App } = require('@octokit/app')
+const { retry } = require('@octokit/plugin-retry')
+const Client = Octokit.plugin(retry)
+const stripAnsi = require('strip-ansi')
+const { shouldCreateIssues, shouldAddCheckToPR } = require('./config')
+const {
+  fmtSummary,
+  getPackageName,
+  getPullNumber,
+  getTestSuiteTitles,
+  parseErrLineNumberFromStack,
+  parseFirstErrFromFlakeBody,
+  parseMandatoryTestIssuesFromPullBody,
+  parsePathFromStack,
+} = require('./utils')
+
+const FlakeLabel = 'FLAKEY'
+const defaults = {
+  owner: process.env.CIRCLE_PROJECT_USERNAME || 'celo-org',
+  repo: process.env.CIRCLE_PROJECT_REPONAME || 'celo-monorepo',
+}
+
+const getLabels = () => {
+  return [FlakeLabel, getPackageName(), process.env.CIRCLE_JOB]
+}
+
+class GitHub {
+  constructor(app, rest) {
+    if (typeof app === 'undefined' || typeof rest === 'undefined') {
+      throw new Error('GitHub constructor should not be called directly. Please use GitHub.build()')
+    }
+    this.app = app
+    this.rest = rest
+  }
+
+  static async build() {
+    const app = new App({
+      id: 71131,
+      privateKey: process.env.FLAKE_TRACKER_SECRET.replace(/\\n/gm, '\n'),
+      //privateKey: privateKey,
+    })
+
+    const rest = await auth(app)
+
+    return new GitHub(app, rest)
+  }
+
+  async renew() {
+    if (typeof this === 'undefined') {
+      throw new Error('GitHub.renew() cannot be called before GitHub.build()')
+    }
+    this.rest = await auth(this.app)
+  }
+
+  async report(flakes, skippedTests) {
+    const promises = []
+
+    if (shouldCreateIssues) {
+      // Check list of ALL flakey issues to ensure no duplicates
+      const knownFlakes = await this.fetchKnownFlakes()
+      const newFlakes = flakes.filter((flake) => !knownFlakes.includes(flake.title))
+      if (newFlakes.length) {
+        promises.push(this.issues(newFlakes))
+      }
+    }
+
+    if (shouldAddCheckToPR) {
+      promises.push(this.endCheck(flakes, skippedTests))
+    }
+
+    console.log('\nSending flake tracker results to GitHub...\n')
+
+    return Promise.all(promises)
+  }
+
+  async issues(flakes) {
+    return Promise.all(flakes.map((f) => this.issue(f)))
```

Comment:
Sounds good. The safeExec function that all the API calls use handles errors with console.error as I didn't want the workflow to fail just because the flake tracker couldn't create an issue. I'll add some comments to clarify